inference-engine/ie_bridges/python/src/openvino/inference_engine/ie_api_impl.cpp

```diff
@@ -665,6 +668,163 @@ void InferenceEnginePython::IECore::addExtension(const std::string& ext_lib_path
+            inDataConfig.push_back(inpConf);
+        }
+
+        // Output shapes
+        for (auto& outShape : outShapes) {
+            InferenceEngine::SizeVector order(outShape.size());
+            std::iota(order.begin(), order.end(), 0);
+
+            InferenceEngine::DataConfig outConf;
+            outConf.desc = InferenceEngine::TensorDesc(InferenceEngine::Precision::FP32, outShape, {outShape, order, offset});
+            outDataConfig.push_back(outConf);
+
+        }
+        InferenceEngine::LayerConfig layerConfig;
+        layerConfig.inConfs = inDataConfig;
+        layerConfig.outConfs = outDataConfig;
+
+        conf.push_back(layerConfig);
+
+        return InferenceEngine::StatusCode::OK;
+    }
+
+    InferenceEngine::StatusCode init(InferenceEngine::LayerConfig& config,
+                                     InferenceEngine::ResponseDesc* resp) noexcept override {
+        return InferenceEngine::OK;
+    }
+
+    InferenceEngine::StatusCode execute(std::vector<InferenceEngine::Blob::Ptr>& inputs,
+                                        std::vector<InferenceEngine::Blob::Ptr>& outputs,
+                                        InferenceEngine::ResponseDesc* resp) noexcept override {
+        _import_array();
+
+        PyObject* args = PyList_New(inputs.size());
+        for (size_t i = 0; i < inputs.size(); ++i) {
+            void* data = inputs[i]->buffer();
+            std::vector<size_t> dims = inputs[i]->getTensorDesc().getDims();
+            std::vector<npy_intp> npy_dims(dims.begin(), dims.end());
+            PyObject* arr = PyArray_SimpleNewFromData(dims.size(), npy_dims.data(), NPY_FLOAT32, data);
+            PyList_SET_ITEM(args, i, arr);
+        }
+        PyObject* res = PyObject_CallMethodObjArgs(impl, PyUnicode_FromString("execute"), args, NULL);
+        Py_DECREF(args);
+
+
+        void* data = PyArray_DATA((PyArrayObject*)res);
+        std::memcpy(outputs[0]->buffer(), data, outputs[0]->byteSize());
+        Py_DECREF(res);
+
+        return InferenceEngine::OK;
+    }
+
+private:
+    std::vector<ngraph::Shape> inpShapes;
+    std::vector<ngraph::Shape> outShapes;
+    PyObject* impl;
+};
+
+class PyExtension : public InferenceEngine::IExtension
+{
+public:
+    explicit PyExtension(PyObject* def) {
+        this->def = def;
+
+        // Get a layer type
+        PyObject* op = PyObject_GetAttrString(def, "op");
+        PyObject* bytes = PyUnicode_AsUTF8String(op);
+        opName = PyBytes_AsString(bytes);
+        Py_DECREF(bytes);
+        Py_DECREF(op);
+    }
+
+    void Unload() noexcept override {}
+    void Release() noexcept override { delete this; }
+    void GetVersion(const InferenceEngine::Version*&) const noexcept override {}
+
+    std::map<std::string, ngraph::OpSet> getOpSets() override {
+        std::map<std::string, ngraph::OpSet> opsets;
+        ngraph::OpSet opset;
+        opset.insert<ov::op::util::FrameworkNode>();
```

Comment:
@dkurt, believe you are going to rebase changes on the latest API 2.0 API eventually. If this is the case, could you consider to avoid using FrameworkNode? FW Node will likely be changed and serves a goal to represent unknown operation from FW in partially converted scenarios, but in your case it is known operation with kernel implementation and we can define all necessary methods that we require from a custom OV operation but in Python this time. I would recommend to discuss the design first with @ilyachur.

But anyway this is an awesome feature!

Reply:
BTW, I've heard that someone also tried Python custom layers for 2022.1 release. Do you know?

Reply:
We have frontend::OpExtension and frontend::ConversionExtension that can be implemented in Python (as well as in C++). This is not really a new custom op in Python, it is just a conversion code that should use already known standard operations or custom operations implemented in C++. But it is more complete in C++ in comparison to Python. For example, only in C++ you can really use your own custom ops to do the conversion from the original FW custom op. In python you can load a library compiled in C++ that uses  custom ops if they are really required. That works with new FrontEnd only, ONNX and Paddle. Don't know about real practical implementations of completely Python custom ops at this time.